grails-app/services/com/unifina/service/CanvasService.groovy

```diff
@@ -201,8 +202,50 @@ class CanvasService {
 		return grailsLinkGenerator.link(controller: 'canvas', action: 'editor', id: canvas.id, absolute: true)
 	}
 
+	@CompileStatic
+	def addExampleCanvases(SecUser user, List<Canvas> examples) {
+		for (Canvas c : examples) {
+			switch (c.exampleType) {
+				// Create a copy of the example canvas for the user and grant all permissions.
+				case ExampleType.COPY:
+					Canvas canvas = new Canvas()
+					canvas.dateCreated = c.dateCreated
+					canvas.lastUpdated = c.lastUpdated
+					canvas.name = c.name
+					canvas.json = c.json
+					canvas.state = Canvas.State.STOPPED
+					canvas.hasExports = c.hasExports
+					canvas.adhoc = c.adhoc
+					canvas.runner = null //c.runner
+					canvas.server = null //c.server
+					canvas.requestUrl = null //c.requestUrl
+					canvas.serialization = null //c.serialization
+					canvas.startedBy = null //c.startedBy
+					canvas.exampleType = ExampleType.NOT_SET
+					c.save(validate: true, failOnError: true)
+					permissionService.systemGrant(user, canvas, Permission.Operation.SHARE)
+
+					Map signalPathMap = (Map) JSON.parse(canvas.json)
+					resetUiChannels(signalPathMap)
+					SignalPathService.ReconstructedResult result = reconstructFrom(signalPathMap, user)
+					result.signalPath.setCanvas(canvas)
+					result.signalPath.getModules().each {
+						if (it instanceof ModuleWithUI) {
+							it.ensureUiChannel()
+						}
+					}
+					result.signalPath.ensureUiChannel()
+					break
+				// Grant read permission to example canvas.
+				case ExampleType.SHARE:
+					permissionService.systemGrant(user, c, Permission.Operation.READ)
+					break
+			}
+		}
+	}
+
 	private boolean hasCanvasPermission(Canvas canvas, SecUser user, Permission.Operation op) {
```

Comment:
One-liner private function seems useless now. Could just as well call `permissionService.check` directly in the places where this function is used.

Reply:
I'd like to leave it like that because it is more readable.

Reply:
Ok, no problem